expr/func_highestlowest.go

```diff
@@ -0,0 +1,79 @@
+package expr
+
+import (
+	"math"
+	"sort"
+
+	"github.com/grafana/metrictank/consolidation"
+	schema "gopkg.in/raintank/schema.v1"
+
+	"github.com/grafana/metrictank/api/models"
+)
+
+type FuncHighestLowest struct {
+	in      GraphiteFunc
+	n       int64
+	fn      string
+	highest bool
+}
+
+func NewHighestLowestConstructor(fn string, highest bool) func() GraphiteFunc {
+	return func() GraphiteFunc {
+		return &FuncHighestLowest{fn: fn, highest: highest}
+	}
+}
+
+func (s *FuncHighestLowest) Signature() ([]Arg, []Arg) {
+	if s.fn != "" {
+		return []Arg{
+			ArgSeriesList{val: &s.in},
+			ArgInt{key: "n", val: &s.n},
+		}, []Arg{ArgSeriesList{}}
+	}
+	return []Arg{
+		ArgSeriesList{val: &s.in},
+		ArgInt{key: "n", val: &s.n},
+		ArgString{key: "func", val: &s.fn, validator: []Validator{IsConsolFunc}},
+	}, []Arg{ArgSeriesList{}}
+}
+
+func (s *FuncHighestLowest) Context(context Context) Context {
+	return context
+}
+
+func (s *FuncHighestLowest) Exec(cache map[Req][]models.Series) ([]models.Series, error) {
+	series, err := s.in.Exec(cache)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(series) == 0 {
+		return series, nil
+	}
+
+	consolidationFunc := consolidation.GetAggFunc(consolidation.FromConsolidateBy(s.fn))
+
+	// Calculates the consolidated value for each series and stores it in a map
+	// where the key is a pointer to the first datapoint
+	consolidationVals := make(map[*schema.Point]float64, len(series))
```

Comment:
why pointers? i would just use the float64 as map key

Reply:
The pointer is to the first Datapoint of the series. (not sure what float64 you're referring to)

Reply:
the float64 of the first Datapoint of the series

Reply:
I'm not sure how that would work.. two series can have equal values in the first datapoint, so it wouldn't work as a map key.